### DISCUSSION TITLE
Polygon Confiner for 2D Phantom Cameras

### DISCUSSION THREAD

**Discussion by LouLou-alive** [ramokz/phantom-camera#237] [Ideas]
I'd really like to have a polygon confiner for 2d phantom cameras similar to [unity's cinemachine confiner](https://docs.unity3d.com/Packages/com.unity.cinemachine@2.8/manual/CinemachineConfiner2D.html) for all sorts of 2D side view games.
Because I need this in my current hobby project I already implemented a primitive but functional version which creates the inner polygon based on _camera_frame_rect and some trigonometry etc. Specifics have to be fledged out though and it might have to be optimised. Here's some sketching of how the camera2D position clamping look's like using the inner polygon.

```gdscript
# "confiner.confined_polygon" is inner Polygon2D
# "was_camera_in_confiner" is for a sticky camera behaviour; if the target/value is outside the confiner, the camera won't follow the target with the closest position on the confiner but rather stick to the edge of the confiner where the target/value left.
func _set_limit_clamp_position(value: Vector2) -> Vector2:
	if not confiner or confiner.confined_polygon.polygon.size() == 0:
		var camera_frame_rect_size: Vector2 = _camera_frame_rect().size
		value.x = clampf(value.x, _limit_sides.x + camera_frame_rect_size.x / 2.0, _limit_sides.z - camera_frame_rect_size.x / 2.0)
		value.y = clampf(value.y, _limit_sides.y + camera_frame_rect_size.y / 2.0, _limit_sides.w - camera_frame_rect_size.y / 2.0)
		return value

	elif Geometry2D.is_point_in_polygon(
		value - confiner.global_position, confiner.confined_polygon.polygon
	):
		# was_camera_in_confiner = true
		return value
	# elif not was_camera_in_confiner:
	# 	return get_global_position()

	var closest_points = {}
	var poly_size = confiner.confined_polygon.polygon.size()
	for i in range(0, poly_size):
		var point = Geometry2D.get_closest_point_to_segment(
			value,
			confiner.global_position + confiner.confined_polygon.polygon[i],
			confiner.global_position + confiner.confined_polygon.polygon[(i + 1) % poly_size]
		)
		closest_points[value.distance_to(point)] = point
	var closest = closest_points[closest_points.keys().min()]
	return closest
```

Maybe the confiner should be an individual node/component, replacing the current limit_node, for a more maintainable design. The confiner component script then extends CollisionPolygon2D (or something similar) for Godot's built in manual vertex editing. The inner polygon is automatically calculated and held as a variable in the script, not editable in the editor. 
This way a Tilemap can be assigned to the confiner node (not directly to the pcam2D) and the outer polygon vertices are calculated automatically with a press of a button for example. This would also take concave Tilemap edges into account.

Also, having the confiner be an individual component is the better choice in terms of performance if there are multiple pcams using the same limits.

Honestly, I'm quite green regarding godot, lmk what you think and if I should attempt to further implement something like this.

**Comment by ramokz** [comment]
I have been toying around with the idea of an addon specific `CollisionShape`/`CollisionPolygon` node that provides functionalities like that. Mainly for better usability and to provide common features out-of-the-box, so that option is definitely on the table.

There is support for using a `CollisionShape` nodes as the `Limit Node Target`, but I'm guessing you're essentially looking to specify the corners and edges of the collision shape using a `CollisionPolygon` as well?

**Comment by A-James** [comment]
Was there any further development of this idea?

As you said, I'm in the exact scenario of wanting to use a `CollisionPolygon2d` rather than a `CollisionShape2d`: My world environment is irregular in shape, so the camera must be limited to an enclosed polygon to move coherently. As the parent comment noted this is implemented in Cinemachine Confiner 2D in Unity, but I've been unable to find any feasible approach in Godot other than rolling from scratch.

In theory a similar effect is also achievable by splitting my map into zones each defined by a `CollisionShape2d`, and writing some logic to change the active `pcam` bounds to a new `CollisionShape2d` as my player moves through them - potentially even manually setting the limit coordinates and interpolating between them for greater control.

Am I missing a more obvious solution?

**Reply by ramokz** [reply]
> Was there any further development of this idea?

Someone made a feature request for it and offered to try implementing it #430.
No updates about any progress as of yet, so can't tell if it's still being worked on.
To your point, also noted in the issue that it is a from-scratch implementation, which I can see being a fair bit of work to both do and get to behave consistently.

> In theory a similar effect is also achievable by splitting my map into zones each defined by a CollisionShape2d, and writing some logic to change the active pcam bounds to a new CollisionShape2d as my player moves through them - potentially even manually setting the limit coordinates and interpolating between them for greater control.

This is what I tend to suggest at this time.
It's a bit DIY by its nature, but it is an out-of-box supported feature and _should_ cover most use-cases.
